samplers/aws/xray/remote_sampler.go

```diff
@@ -0,0 +1,219 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xray
+
+import (
+	"context"
+	crypto "crypto/rand"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+// global variable for logging
+var globalLogger logr.Logger
+
+// remoteSampler is a sampler for AWS X-Ray which polls sampling rules and sampling targets
+// to make a sampling decision based on rules set by users on AWS X-Ray console
+type remoteSampler struct {
+	// List of known centralized sampling rules.
+	manifest *manifest
+
+	// xrayClient is used for getting quotas and sampling rules.
+	xrayClient *xrayClient
+
+	// pollerStart, if true represents rule and target pollers are started.
+	pollerStart bool
+
+	// samplingRulesPollingInterval, default is 300 seconds.
+	samplingRulesPollingInterval time.Duration
+
+	// Unique ID used by XRay service to identify this client.
+	clientID string
+
+	// Provides system time.
+	clock clock
+
+	mu sync.RWMutex
+}
+
+// Compile time assertion that remoteSampler implements the Sampler interface.
+var _ sdktrace.Sampler = (*remoteSampler)(nil)
+
+// NewRemoteSampler returns a sampler which decides to sample a given request or not
+// based on the sampling rules set by users on AWS X-Ray console. Sampler also periodically polls
+// sampling rules and sampling targets.
+func NewRemoteSampler(ctx context.Context, opts ...Option) (sdktrace.Sampler, error) {
+	cfg := newConfig(opts...)
+
+	// Generate clientID
+	var r [12]byte
+
+	_, err := crypto.Read(r[:])
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate client ID: %w", err)
+	}
+
+	id := fmt.Sprintf("%02x", r)
+
+	clock := &defaultClock{}
+
+	m := &manifest{
+		rules: []*rule{},
+		index: map[string]*rule{},
+		clock: clock,
+	}
+
+	remoteSampler := &remoteSampler{
+		clock:                        clock,
+		manifest:                     m,
+		clientID:                     id,
+		xrayClient:                   newClient(cfg.endpoint),
+		samplingRulesPollingInterval: cfg.samplingRulesPollingInterval,
+	}
+
+	// starts the rule poller
+	remoteSampler.start(ctx)
+
+	return remoteSampler, nil
+}
+
+func (rs *remoteSampler) ShouldSample(parameters sdktrace.SamplingParameters) sdktrace.SamplingResult {
+	// ToDo: add business logic for remote sampling
+
+	return sdktrace.TraceIDRatioBased(0.05).ShouldSample(parameters)
+}
+
+func (rs *remoteSampler) Description() string {
+	return "remote sampling with AWS X-Ray"
```

Comment:
This needs to be the same across languages and a programmatic string I think. I don't think we've spec'd this well yet but can probably be a bit closer to this for now

https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsXrayRemoteSampler.java#L116